### PR TITLE
[JN-814] Multiple Combobox questions can fetch choices by path

### PR DESCRIPTION
**Pull request by MatthewBemis**
#### DESCRIPTION

This change allows a study admin to upload a JSON string list containing dropdown choices and reference the path in a `multiplecombox` question. `multiplecombox` is sort of a hidden feature. It was originally added in order to support the Medications question type, which has a dropdown with almost 3,000 choices. The normal React-select dropdown and SurveyJS Tagbox don't perform well with that many options, so this was a custom question type that supports virtualization and improves performance. HeartHive's list allegedly has 15,000 medications, so it was necessary to use the virtualized combobox approach.

https://github.com/broadinstitute/juniper/assets/7257391/3af2a230-327a-488a-8e9b-96cb4d7a8fac

#### TO TEST:  *(simple manual steps for confirming core behavior -- used for pre-release checks)*

* Reboot AdminApiApp and ParticipantApiApp
* Make sure your UIs are running and have the latest ui-core changes from this PR
* Repopulate OurHealth
* Go to the manage media page: https://localhost:3000/ourhealth/studies/ourheart/env/sandbox/media
* Upload the attached colors.json file: [colors.json](https://github.com/broadinstitute/juniper/files/14057383/colors.json)
* Click the "Copy path" button in the Site Media list
* Edit the Basics survey and using the JSON Editor, add the following question (you may have to adjust the path depending on where you're testing it):
```
{
    "type": "multiplecombobox",
    "name": "oh_oh_favColor",
    "placeholder": "Select a color",
    "title": "Please choose your favorite color(s)",
    "description": "",
    "isRequired": false,
    "choicesByUrl": {
        "url": "/api/public/portals/v1/ourhealth/env/live/siteImages/2/colors.json"
    }
}
```
* In the preview mode, verify that it loads the color options
* Save the survey

Test from the participant side:
* Log in as a new user (so they'll get the updated survey)
* Complete the basics survey, and select some colors
* Go back to the admin app and verify that you see the users selections in their profile

### CODE REVIEW

File: ui-admin/src/portal/images/SiteImageList.tsx [hunk 1]
```diff
@@ -74,6 +74,17 @@ export default function SiteImageList({ portalContext, portalEnv }:
...
+  }, {
+    header: '',
+    id: 'copy',
+    cell: ({ row: { original: image } }) => <button onClick={() => {
+      navigator.clipboard.writeText(getImageUrl(portalContext.portal.shortcode,
+        image.cleanFileName,
+        image.version))
+    }}
+    className="btn btn-secondary">
+      <FontAwesomeIcon icon={faClipboard} className="fa-lg"/> Copy Path
```
Comment (devonbush) [inline ui-admin/src/portal/images/SiteImageList.tsx:86]: maybe "Copy Url" would be a better name?

File: ui-admin/src/portal/images/SiteImageUploadModal.tsx [hunk 2]
```diff
@@ -11,7 +11,7 @@ import { useFileUploadButton } from 'util/uploadUtils'
 
 
 export const allowedImageTypes = ['gif', 'ico', 'jpeg', 'jpg', 'png', 'svg', 'webp']
-export const allowedDocumentTypes = ['pdf']
+export const allowedDocumentTypes = ['pdf', 'json']
```
Comment (devonbush) [inline ui-admin/src/portal/images/SiteImageUploadModal.tsx:14]: let's go ahead and add .txt and .csv here -- I'm sure we'll want them eventually

File: ui-core/src/components/MultipleCombobox.tsx [hunk 3]
```diff
@@ -3,22 +3,44 @@ import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
...
   id?: string
   initialValue?: ComboboxItem[]
   itemToString: (item: ComboboxItem) => string
   options: ComboboxItem[]
+  choicesByUrl?: string
   placeholder?: string
   onChange?: (value: ComboboxItem[]) => void
 }
 
 // TODO: Add JSDoc
```
Comment (devonbush) [inline ui-core/src/components/MultipleCombobox.tsx]: let's fix this todo now

File: ui-core/src/surveyjs/medications-question.ts [hunk 4]
```diff
@@ -6,7 +6,7 @@ ComponentCollection.Instance.add({
   name: 'medications',
   title: 'Medications',
   questionJSON: {
-    type: 'multiple-combobox',
+    type: 'multiplecombobox',
```
Comment (devonbush) [inline ui-core/src/surveyjs/medications-question.ts:9]: what's the reason for this change?

**Review by devonbush — APPROVED** [review]
nice!  A couple of questions and some cleanup suggestions

File: ui-admin/src/portal/images/SiteImageList.tsx (hunk 1 above)
Comment (MatthewBemis) [reply ui-admin/src/portal/images/SiteImageList.tsx:86]: As-implemented, it's just the path. My thinking was that using the path makes the survey JSON a bit more portable across demo and prod. Since the files are publicly accessible, I wanted to avoid bleeding across deployments.

File: ui-core/src/surveyjs/medications-question.ts (hunk 4 above)
Comment (MatthewBemis) [reply ui-core/src/surveyjs/medications-question.ts:9]: I had originally started to go down the path of making MultipleCombobox a selectable question type in the question designer, and the dash was causing problems for some of the templating we use for question types. I left it like this because it's consistent with other question type naming and will also make our lives slightly easier if we ever introduce it as a visible feature.

File: ui-core/src/components/MultipleCombobox.tsx [hunk 5]
```diff
@@ -3,22 +3,46 @@ import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
...
 import { useCombobox, useMultipleSelection } from 'downshift'
-import React, { useCallback, useMemo, useRef, useState } from 'react'
+import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react'
 import { trimStart } from 'lodash'
 
 type MultipleComboboxProps<ComboboxItem> = {
   id?: string
   initialValue?: ComboboxItem[]
   itemToString: (item: ComboboxItem) => string
   options: ComboboxItem[]
```
Comment (connorlbark) [inline ui-core/src/components/MultipleCombobox.tsx:13]: should options be optional now that `choicesByUrl` exists? and perhaps `choicesByUrl` should be `optionsByUrl` for consistency?

**Review by connorlbark — APPROVED** [review]
worked perfectly for me! outside of the scope, but I wasn't sure about the styling of the combobox (specifically how it highlighted blue on click), but what you added here looks great!
